useLiteralContent data validation

maxsam4 opened the pull request:
### Description
- Added datatype check for `settings.metadata.useLiteralContent`
- Fixed #5456 

### Checklist
- [X] Code compiles correctly
- [X] All tests are passing
- [X] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

stackenbotten commented:
There was an error when running `test_check_style` for commit `5a7cf08db3027d75081d2e698dd51b807bef5abb`:
```
Error: Format error for if/for:
 libsolidity/interface/StandardCompiler.cpp:581: if(settings.isMember("metadata")) {
 libsolidity/interface/StandardCompiler.cpp:581: if(settings.isMember("metadata")) {

```
Please check that your changes are working as intended.

leonardoalt commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -298,6 +298,8 @@ boost::optional<Json::Value> checkOptimizerKeys(Json::Value const& _input)
 boost::optional<Json::Value> checkMetadataKeys(Json::Value const& _input)
 {
+	if (_input.isObject() && _input.isMember("useLiteralContent") && !_input["useLiteralContent"].isBool())
+		return formatFatalError("JSONError", "\"settings.metadata.useLiteralContent\" must be a bool");
```
```suggestion
		return formatFatalError("JSONError", "\"settings.metadata.useLiteralContent\" must be Boolean");
```

chriseth commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -576,12 +578,13 @@ Json::Value StandardCompiler::compileInternal(Json::Value const& _input)
-		return *result;
-
-	m_compilerStack.useMetadataLiteralSources(metadataSettings.get("useLiteralContent", Json::Value(false)).asBool());
+	if (settings.isMember("metadata"))
```
The style of the other lines suggests that this check is not needed, or is it?

leonardoalt replied:
I had the same question.

maxsam4 replied:
Optional objects are using this check. For example, `if (settings.isMember("optimizer"))` is present on L514 so I added this for consistency. That being said, it was working earlier as well when `metadata` was absent so we probably don't need the check.

maxsam4 replied:
Everything is working fine without this check as well. Let me know if you want to keep the check or remove it. The check can be a minor optimization if `settings.metadata` is mostly missing. Also makes it more explicit that `settings.optimizer` is optional.

chriseth replied:
Please remove the check

maxsam4 replied:
Removed.

chriseth commented on test/cmdlineTests/standard_missing_type_useLiteralContent/input.json:
```diff
@@ -0,0 +1,19 @@
+{
```
Shouldn't the file be named "standard_missing_key_..."?

maxsam4 replied:
Yep, changed the name.

chriseth reviewed (changes requested):
Nice, that's almost it already!

leonardoalt commented:
Thanks @maxsam4 !

@chriseth Didn't we have a PR that validated every field?

leonardoalt commented:
Looks good! Could you please squash the commits?

maxsam4 committed:
useLiteralContent data validation
Squashed commit of the following:

commit bbceee6997c1b951eb6054f7d5b7560fe1773690
Author: Mudit Gupta <guptamudit@ymail.com>
Date:   Tue Jan 22 23:27:49 2019 +0530

    Removed extra check

commit 4f7c3cc58e94836b1dddcc5f2938da14adad6252
Author: Mudit Gupta <guptamudit@ymail.com>
Date:   Tue Jan 22 17:51:05 2019 +0530

    Updated test error message

commit 5da45f2a23840e7bb1978853e1c184faec65b3e0
Author: Mudit Gupta <guptamudit@ymail.com>
Date:   Tue Jan 22 17:46:09 2019 +0530

    Renamed test

commit e661418deac25f31c49091be3c80244f566fc14c
Author: Leonardo <leo@ethereum.org>
Date:   Tue Jan 22 17:41:56 2019 +0530

    Update libsolidity/interface/StandardCompiler.cpp

    Co-Authored-By: maxsam4 <guptamudit@ymail.com>

commit d289b4dc1956ab736aada613e810cf2e2d124d4d
Author: Mudit Gupta <guptamudit@ymail.com>
Date:   Tue Jan 22 17:29:40 2019 +0530

    changed style

commit 5a7cf08db3027d75081d2e698dd51b807bef5abb
Author: Mudit Gupta <guptamudit@ymail.com>
Date:   Tue Jan 22 17:20:37 2019 +0530

    Added test case for missing useLiteralContent

commit 6e866c46aa92d6a89f2b341bd717c3886946f3ff
Author: Mudit Gupta <guptamudit@ymail.com>
Date:   Tue Jan 22 17:09:30 2019 +0530

    input check for useLiteralContent

maxsam4 commented:
> Looks good! Could you please squash the commits?

Thanks. 
Just force pushed a squashed commit with the default squash message from git.

chriseth reviewed (approved):
Thanks a lot!

maxsam4 commented:
It was a pleasure working on this. I'll definitely pick up more tasks in my free time :)
